virtcontainers/kata_agent.go

```diff
@@ -584,6 +584,33 @@ func (k *kataAgent) setProxy(sandbox *Sandbox, proxy proxy, pid int, url string)
 	return nil
 }
 
+func (k *kataAgent) checkAgent(sandbox *Sandbox) error {
+	savedCtx := k.ctx
+	ctx, cancel := context.WithTimeout(k.ctx, checkRequestTimeout)
+	k.ctx = ctx
+	defer func() {
+		cancel()
+		k.ctx = savedCtx
+	}()
+
+	// wait agent is running
+	switch k.vmSocket.(type) {
+	case Socket:
+		if err := sandbox.hypervisor.waitAgent(ctx); err != nil {
```

Comment:
By waiting for the agent only in case of `Socket`, which means it does not apply for VSOCK, we assume the agent knows about the hypervisor implementation. This is not okay, because that's the whole point about the hypervisor interface (to provide some abstraction). The whole timeout handling + decision to return immediately if the type is `Socket` should be handled by the hypervisor implementation directly.
@bergwolf WDYT?

Reply:
Any thoughts on this @bergwolf?

Reply:
It's worth pointing out here that the grpc client (used by `k.check()` below) will keep dialing the grpc server running in the agent until it gets a response.